### PR TITLE
Rewrite entity pipe character in JsonEntityId

### PR DESCRIPTION
**Pull request by mariussoutier**
Unfortunately cannot test if entity actually starts up. `entityRef.tell(Command())` also works when name is illegal.

### CODE REVIEW

File: akka-components/src/main/scala/ClusterComponent.scala [hunk 1]
```diff
@@ -415,11 +415,17 @@ object ClusterComponent {
     trait JsonEntityId {
       _: ShardedT =>
 
+      /**
+       * The placeholder that will replace any | in the produced String as this character is illegal due to Akka's entityId encoding.
+       * When overriding keep the new placeholder permanently or existing entities won't be resolved.
+       * */
+      val barPlaceholder = "❘❘bar❘❘"
```
Comment (mariussoutier) [inline akka-components/src/main/scala/ClusterComponent.scala:422]: These are UTF-8 bars

Comment (an-tex) [reply akka-components/src/main/scala/ClusterComponent.scala:422]: interesting approach. this comment might be useful in the code comment too (as one might wonder why a single | is an issue but a || isn't (as they look very similar)

File: akka-components/src/test/scala/ClusterComponentSpec.scala [hunk 2]
```diff
@@ -255,7 +255,42 @@ class ClusterComponentSpec extends ScalaTestWithActorTestKit(ConfigFactory.parse
 
         val component = new ComponentObject.Component
         ComponentObject.init(component).delayedInit()
-        component.entityRefFor(ComponentObject.EntityId("a", "b"))
+        val entityRef = component.entityRefFor(ComponentObject.EntityId("a", "b"))
+        entityRef.entityId shouldBe """{"id1":"a","id2":"b"}"""
+
+        val entityRefWithPlaceholder = component.entityRefFor(ComponentObject.EntityId("a|b", "c|d"))
+        entityRefWithPlaceholder.entityId shouldBe """{"id1":"a❘❘bar❘❘b","id2":"c❘❘bar❘❘d"}"""
```
Comment (an-tex) [inline akka-components/src/test/scala/ClusterComponentSpec.scala:262]: this tests only the serialization? testing the deserialization would be great too

**Review by an-tex — CHANGES_REQUESTED** [review]

File: akka-components/src/main/scala/ClusterComponent.scala (hunk 1 above)
Comment (mariussoutier) [reply akka-components/src/main/scala/ClusterComponent.scala:422]: added

File: akka-components/src/test/scala/ClusterComponentSpec.scala (hunk 2 above)
Comment (mariussoutier) [reply akka-components/src/test/scala/ClusterComponentSpec.scala]: added codec tests

**Comment by an-tex** [comment]
> > Unfortunately cannot test if entity actually starts up. entityRef.tell(Command()) also works when name is illegal.
> 
> did you try an EventSourced entity? I think this issue only arises there (when trying to recover)

@mariussoutier in this case it's easiest to test it using the  `net.sc8s.akka.components.testkit.ClusterComponentTestKitSpec` in the `akka-components-testkit` module. I've added an example in https://github.com/an-tex/sc8s/pull/411/commits/bd925e1370180768384a58a16d7d389ba9d0e6cd feel free to change/extend it

**Comment by mariussoutier** [comment]
> I've added an example in [bd925e1](https://github.com/an-tex/sc8s/commit/bd925e1370180768384a58a16d7d389ba9d0e6cd) feel free to change/extend it

I've played around with it and I think it's enough to test it.

**Comment by an-tex** [comment]
thx @mariussoutier !

**Pull request merged**